ethereum/core/src/test/java/org/hyperledger/besu/ethereum/chain/BadBlockManagerTest.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Hyperledger Besu Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+package org.hyperledger.besu.ethereum.chain;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.core.BlockchainSetupUtil;
+import org.hyperledger.besu.ethereum.trie.MerkleTrieException;
+import org.hyperledger.besu.plugin.services.exception.StorageException;
+
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class BadBlockManagerTest {
+
+  final BlockchainSetupUtil chainUtil = BlockchainSetupUtil.forMainnet();
+  final Block block = chainUtil.getBlock(1);
+  final BadBlockManager badBlockManager = new BadBlockManager();
+
+  public static Stream<Arguments> getInternalExceptions() {
+    return Stream.of(
+        Arguments.of("StorageException", new StorageException("oops")),
+        Arguments.of("MerkleTrieException", new MerkleTrieException("fail")));
+  }
+
+  @Test
+  public void addBadBlock_addsBlock() {
+    BadBlockManager badBlockManager = new BadBlockManager();
+    final BadBlockCause cause = BadBlockCause.fromValidationFailure("failed");
+    badBlockManager.addBadBlock(block, cause);
+
+    assertThat(badBlockManager.getBadBlocks()).containsExactly(block);
+  }
+
+  @ParameterizedTest(name = "[{index}] {0}")
+  @MethodSource("getInternalExceptions")
+  public void addBadBlock_ignoresInternalError(final String caseName, final Exception err) {
+    BadBlockManager badBlockManager = new BadBlockManager();
+    final BadBlockCause cause = BadBlockCause.fromProcessingError(err);
+    badBlockManager.addBadBlock(block, cause);
+
+    assertThat(badBlockManager.getBadBlocks()).isEmpty();
+  }
+
+  @Test
+  public void addBadBlock_doesNotIgnoreRuntimeException() {
```

Comment:
This could be a problem. If a runtime problem is repeatable, and occurs for any out-of-protocol reason/violation, we will get a chain halt. The CL will continue to try to add the block, as it is canonical, and besu will refuse since it is already present in the bad block manager. We ran into this a lot building out v3 engine apis. I don't know how to better answer the question "did this RuntimeException reflect a protocol violation".

Reply:
100% Agree.  I have a follow-up PR coming to fix this.  I wanted to isolate behavior changes in a separate PR.

Reply:
To be clear - this is a test validating the current behavior (which was previously untested).  I agree this behavior is problematic and should be fixed.  I will push fixes / functional changes in a follow-up PR.

Reply:
Here is the follow-up: https://github.com/hyperledger/besu/pull/6678